### PR TITLE
Add one to one relation between SiteSettings and django.contrib.sites.Site

### LINKED ISSUE #1213: Merge Site and SiteSettings models
Currently fields site domain and name are stored in two models.
- django.contrib.sites.models.Site
- saleor.site.models.SiteSettings
We should use only django Site framework to get those data.

### PR DESCRIPTION
**Pull request by koradon**
ToDo:

- [x] one 2 one relation between Site and SiteSettings
- [x] data migration from SiteSettings to Site
- [x] remove fields name and domain from SiteSettings
- [x] update dashboard/sites/form
- [x] monkey patch for sites.model.SiteManager.get_current to always fetch data from our thread safe cache

Fixes #1213 #1195

### CODE REVIEW

File: saleor/core/middleware.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import logging
 
 from django.conf import settings
+from django.contrib.sites import models as sites
```
Comment (elwoodxblues) [inline saleor/core/middleware.py]: Why `import models as sites`? Can't we just explicitly import what we need in this module? From what I see we use only SiteManager here, we can import only that.

File: saleor/core/utils/__init__.py [hunk 2]
```diff
@@ -46,7 +46,7 @@ def build_absolute_uri(location, is_secure=False, site_settings=None):
     # type: (str, bool, saleor.site.models.SiteSettings) -> str
     if site_settings is None:
         site_settings = get_site_settings()
-    host = site_settings.domain
+    host = site_settings.site.domain
```
Comment (elwoodxblues) [inline saleor/core/utils/__init__.py]: You can get Site directly from contrib.sites

File: saleor/dashboard/order/utils.py [hunk 3]
```diff
@@ -10,10 +10,10 @@
 
 
 def get_statics_absolute_url(request):
-    site = get_site_settings_from_request(request)
+    site_settings = get_site_settings_from_request(request)
```
Comment (elwoodxblues) [inline saleor/dashboard/order/utils.py]: In this function you need to get the domain from Site model. You can fetch the Site instance directly, without accessing it through `site_settings`.

File: saleor/dashboard/sites/views.py [hunk 4]
```diff
@@ -17,16 +21,24 @@ def index(request):
...
     formset = AuthorizationKeyFormSet(
         request.POST or None, queryset=authorization_qs,
-        initial=[{'site_settings': site}])
-    if all([form.is_valid(), formset.is_valid()]):
-        site = form.save()
+        initial=[{'site_settings': site_settings}])
+    if all([site_settings_form.is_valid(),
+            formset.is_valid()]):
+        site = site_form.save()
+        site_settings.site = site
```
Comment (elwoodxblues) [inline saleor/dashboard/sites/views.py]: `site_settings_form.instance.site = site`

File: saleor/dashboard/sites/views.py [hunk 5]
```diff
@@ -17,16 +21,24 @@ def index(request):
...
+    site_form = SiteForm(request.POST or None, instance=site)
+    authorization_qs = AuthorizationKey.objects.filter(
+        site_settings=site_settings)
     formset = AuthorizationKeyFormSet(
         request.POST or None, queryset=authorization_qs,
-        initial=[{'site_settings': site}])
-    if all([form.is_valid(), formset.is_valid()]):
-        site = form.save()
+        initial=[{'site_settings': site_settings}])
+    if all([site_settings_form.is_valid(),
```
Comment (elwoodxblues) [inline saleor/dashboard/sites/views.py]: site_form.is_valid() missing

File: saleor/dashboard/sites/views.py (hunk 4 above)
Comment (koradon) [reply saleor/dashboard/sites/views.py]: Done

File: saleor/dashboard/sites/views.py (hunk 5 above)
Comment (koradon) [reply saleor/dashboard/sites/views.py]: Done

File: saleor/dashboard/order/utils.py (hunk 3 above)
Comment (koradon) [reply saleor/dashboard/order/utils.py]: Done

File: saleor/core/utils/__init__.py (hunk 2 above)
Comment (koradon) [reply saleor/core/utils/__init__.py]: Done

File: saleor/core/middleware.py (hunk 1 above)
Comment (koradon) [reply saleor/core/middleware.py]: This is not done yet but basicly yes, this will be changed

File: saleor/settings.py [hunk 6]
```diff
@@ -378,7 +380,7 @@ def get_host():
         PROJECT_ROOT, 'saleor', 'static', 'schema.json')
 }
 
-SITE_SETTINGS_ID = 1
+# SITE_SETTINGS_ID = 1
```
Comment (elwoodxblues) [inline saleor/settings.py]: Commented out code

File: saleor/site/migrations/0007_auto_20171027_0856.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+
+def link_to_sites(apps, schema_editor):
+    SiteSettings = apps.get_model('site', 'SiteSettings')
+    Site = apps.get_model('sites', 'Site')
+
+    for site in SiteSettings.objects.all():
```
Comment (elwoodxblues) [inline saleor/site/migrations/0007_auto_20171027_0856.py]: `for settings in SiteSettings.objects.all()` - we are iterating over settings objects not sites.

File: saleor/site/migrations/0007_auto_20171027_0856.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        site.site = Site.objects.get_or_create(domain=site.domain, name=site.name)[0]
+        site.save()
+
```
Comment (elwoodxblues) [inline saleor/site/migrations/0007_auto_20171027_0856.py]: We should also iteratate over existing `Sites` and create appropriate `SiteSettings` object if it doesn't exist.

File: saleor/site/migrations/0007_auto_20171027_0856.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('sites', '0002_alter_domain_unique'),
+        ('site', '0006_auto_20171025_0454'),
+    ]
+
+    operations = [
+        migrations.RunPython(link_to_sites),
```
Comment (elwoodxblues) [inline saleor/site/migrations/0007_auto_20171027_0856.py]: Missing reverse operation. `migrations.RunPython.noop` would be fine.

File: saleor/site/models.py [hunk 10]
```diff
@@ -1,26 +1,83 @@
...
+        THREADED_SITE_CACHE = {}
+
+
+def new_get_by_natural_key(self, domain):
+    return self.prefetch_related('settings').filter(domain__iexact=domain)[0]
+
+
+SiteManager.get_current = new_get_current
+SiteManager.clear_cache = new_clear_cache
+SiteManager.get_by_natural_key = new_get_by_natural_key
```
Comment (elwoodxblues) [inline saleor/site/models.py]: I also think all of the logic related to patching `django.contrib.sites` should be enclosed in separate file (example `patch_sites.py`) or clarity, and the three lines above that do the actual patching also could live in a function `patch_contrib_sites`. In this module we should have some commentary explaining why we're doing this.

File: saleor/dashboard/sites/forms.py [hunk 11]
```diff
@@ -1,12 +1,19 @@
...
+    class Meta:
+        model = Site
+        exclude = []
+
+
 class SiteSettingForm(forms.ModelForm):
     class Meta:
         model = SiteSettings
-        exclude = []
+        exclude = ['site']
```
Comment (bogdal) [inline saleor/dashboard/sites/forms.py]: I'm not a big fan of `exclude`. Can we use `fields` instead? It's safer and much more readable.

Comment (elwoodxblues) [reply saleor/dashboard/sites/forms.py]: I think it's okay in this case. In this form we want to edit all fields of the SiteSettings model except this one. I think it's a more general discussion about which practice is better- to use `exclude` or `fields`. If we agreed on one particular way of doing this, we should refactor each form in one separate pull request. @bogdal Please open a ticket if you think that this should be discussed.

**Pull request merged**